### PR TITLE
onl: build kernel modules within onl

### PR DESCRIPTION
**Pull request by KanjiMonster**
Build kernel modules taken from ONL as part of the onl package build, instead of being split out as individual packages.

There are three kinds of modules in the ONL tree, common ones (like optoe), vendor common ones (shared between several machines), and platform specific ones, so add support for building all three, with the latter based on the enabled platforms.

Since we have quite a few build fixes for these modules during the time, add those as patches before enabling modules builds for the platforms, then convert them one by one by enabling module build and dropping the individual module recipes.

After having converted all platforms we can drop the optional module flag, and enforce the build always. This allows us to significantly reduce the amount of (copied) code, and the effort for adding support for new platforms.

Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>

### CODE REVIEW

**Comment by ideaship** [comment]
Are there any dependencies that I should know of? I don't seem to be able to build this branch without errors (tried as5835 and as7726).

**Comment by KanjiMonster** [comment]
> Are there any dependencies that I should know of? I don't seem to be able to build this branch without errors (tried as5835 and as7726).

There are no dependencies. What kind of errors do you get? Please post them.

**Comment by ideaship** [comment]
> There are no dependencies. What kind of errors do you get? Please post them.

```
Aug 17 08:21:03 |   LD [M]  /tmp/tmp.HKOe7oZ2r4/x86-64-accton-as4630-54pe-cpld.ko
Aug 17 08:21:03 | make[1]: Leaving directory '/tmp/accton-as7726-32x/work-shared/accton-as7726-32x/kernel-build-artifacts'
Aug 17 08:21:03 | make[1]: Entering directory '/tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd3
72b724-r0/git/packages/platforms/accton/x86-64/as4630-54pe/modules/builds'
Aug 17 08:21:03 | make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Aug 17 08:21:03 |   INSTALL /tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd372b724-r0/git/packa
ges/platforms/accton/x86-64/as4630-54pe/modules/builds/lib/modules/5.15.56-yocto-standard-onl/onl/accton/x86-64-accton-as4630-54pe/x86-64-accton-as46
30-54pe-cpld.ko
Aug 17 08:21:03 |   DEPMOD  /tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd372b724-r0/git/packa
ges/platforms/accton/x86-64/as4630-54pe/modules/builds/lib/modules/5.15.56-yocto-standard-onl
Aug 17 08:21:03 | Warning: modules_install: missing 'System.map' file. Skipping depmod.
Aug 17 08:21:03 | make[1]: Leaving directory '/tmp/accton-as7726-32x/work-shared/accton-as7726-32x/kernel-build-artifacts'
Aug 17 08:21:03 | make[1]: Entering directory '/tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd3
72b724-r0/git/packages/platforms/accton/x86-64/as4630-54pe/modules/builds'
Aug 17 08:21:03 | make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Aug 17 08:21:03 |   CC [M]  /tmp/tmp.44jxf0wodF/x86-64-accton-as4630-54pe-leds.o
Aug 17 08:21:03 |   MODPOST /tmp/tmp.44jxf0wodF/Module.symvers
Aug 17 08:21:03 | ERROR: modpost: "as4630_54pe_cpld_write" [/tmp/tmp.44jxf0wodF/x86-64-accton-as4630-54pe-leds.ko] undefined!
Aug 17 08:21:03 | ERROR: modpost: "as4630_54pe_cpld_read" [/tmp/tmp.44jxf0wodF/x86-64-accton-as4630-54pe-leds.ko] undefined!
Aug 17 08:21:03 | make[2]: *** [/tmp/accton-as7726-32x/work-shared/accton-as7726-32x/kernel-source/scripts/Makefile.modpost:134: /tmp/tmp.44jxf0wodF/Module.symvers] Error 1
Aug 17 08:21:03 | make[2]: *** Deleting file '/tmp/tmp.44jxf0wodF/Module.symvers'
Aug 17 08:21:03 | make[1]: *** [/tmp/accton-as7726-32x/work-shared/accton-as7726-32x/kernel-source/Makefile:1801: modules] Error 2
Aug 17 08:21:03 | make[1]: Leaving directory '/tmp/accton-as7726-32x/work-shared/accton-as7726-32x/kernel-build-artifacts'
Aug 17 08:21:03 | make: *** [/tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd372b724-r0/git/make/kmodule.mk:29: modules] Error 2
Aug 17 08:21:03 | make: Leaving directory '/tmp/accton-as7726-32x/work/accton_as7726_32x-poky-linux/onl/1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd372b724-r0/git/packages/platforms/accton/x86-64/as4630-54pe/modules/builds'
Aug 17 08:21:03 | ERROR: oe_runmake failed
Aug 17 08:21:03 | WARNING: exit code 1 from a shell command.
Aug 17 08:21:03 NOTE: recipe onl-1.1+gitAUTOINC+dd61cc1792_8621a0bab7_4bd372b724-r0: task do_compile: Failed
Aug 17 08:21:03 NOTE: recipe keepalived-2.2.2-r0: task do_prepare_recipe_sysroot: Succeeded
Aug 17 08:21:08 NOTE: recipe net-snmp-5.9.1-r0: task do_package: Succeeded
Aug 17 08:21:09 NOTE: Tasks Summary: Attempted 5752 tasks of which 5091 didn't need to be rerun and 1 failed.
Aug 17 08:21:09 
Aug 17 08:21:09 Summary: 1 task failed:
Aug 17 08:21:09   /home/builder/poky/meta-open-network-linux/recipes-extended/onl/onl_git.bb:do_compile
Aug 17 08:21:09 Summary: There were 20 WARNING messages.
Aug 17 08:21:09 Summary: There were 2 ERROR messages, returning a non-zero exit code.
```

**Comment by KanjiMonster** [comment]
Weird. It does build fine for me. Anyway, I added a potential fix, please try now.

**Comment by ideaship** [comment]
The potential fix did not fix it, but dropping all cached build data did. I think you can drop the last commit.

**Comment by ideaship** [comment]
Looks like I drew the wrong conclusion from my observations. I needed to both drop the cache _and_ your fix to build the as5835 (which is particular in that there are symbol references across external modules):

```
Aug 17 17:08:48 | ERROR: modpost: "as4630_54pe_cpld_write" [/tmp/tmp.9KCm115DVM/x86-64-accton-as4630-54pe-leds.ko] undefined!
Aug 17 17:08:48 | ERROR: modpost: "as4630_54pe_cpld_read" [/tmp/tmp.9KCm115DVM/x86-64-accton-as4630-54pe-leds.ko] undefined!
```

**Comment by ideaship** [comment]
I notice that the commits "onl: add support for building common vendor module" and "onl: always build platform modules" change when the commit "onl: build common kernel modules within onl" adds 0011-kmodbuild.sh-don-t-treat-undefined-symbols-as-errors.patch, but the commit messages for all three remain identical.

The commit "onl: build common kernel modules within onl" would benefit from being split into parts:

- remove emerson700 module which didn't build anyway
- add kmodbuild.sh-don-t-treat-undefined-symbols-as-errors.patch (which should not break anything because it just turns an error into a warning)
- remaining changes

The commit messages for  "onl: add support for building common vendor module" and "onl: always build platform modules" should probably mention that they need KBUILD_MODPOST_WARN set, but this is taken care of by kmodbuild.sh-don-t-treat-undefined-symbols-as-errors.patch.

File: recipes-extended/onl/onl_git.bb [hunk 1]
```diff
@@ -62,3 +80,7 @@ ONL_PLATFORM_SUPPORT:x86-64 = " \
     x86-64-delta-ag5648-r0 \
     x86-64-delta-ag7648-r0 \
 "
+
+ONL_MODULE_VENDORS:x86-64 = " \
```
Comment (ideaship) [inline recipes-extended/onl/onl_git.bb:84]: An empty ONL_MODULE_VENDORS should be created in the commit "onl: add support for building common vendor modules" already. Otherwise, the commit does not show how and where the newly introduced variable is meant to be used, and when it does get used in "agema-ag5648: build platform modules via ONL", there is no indication where the variable added to onl_git.bb suddenly came from.

Also, there should be a brief comment above `ONL_MODULE_VENDORS:x86-64 =` indicating what the variable is good for, because from the variable name it is not obvious at all.

**Review by ideaship — CHANGES_REQUESTED** [review]
The commit message for "onl: add module build fixes from our tree" should mention that it is only adding _missing_ build fixes, because many if not most patches have been added (even without the drivers being built) long ago.

File: recipes-extended/onl/onl_git.bb [hunk 2]
```diff
@@ -62,3 +81,11 @@ ONL_PLATFORM_SUPPORT:x86-64 = " \
     x86-64-delta-ag5648-r0 \
     x86-64-delta-ag7648-r0 \
 "
+
+# common vendor kernel modules vendors to build
```
Comment (ideaship) [inline recipes-extended/onl/onl_git.bb:85]: I suppose this was meant to be:
```suggestion
# common vendor kernel modules to build
```

**Review by ideaship — CHANGES_REQUESTED** [review]
LGTM except for a trivial issue.

File: recipes-extended/onl/onl_git.bb (hunk 2 above)
Comment (KanjiMonster) [reply recipes-extended/onl/onl_git.bb:85]: No, that was intentional. This is the list of vendors for which we build the (common) vendor modules (onl package is named `onl-vendor-${vendor}-modules`).

Comment (ideaship) [reply recipes-extended/onl/onl_git.bb:85]: Alright, I understand what you mean, but it's still weird. We are building modules, not vendors, but the comment suggests otherwise. Can you rephrase the comment in a way that is easier to understand?

Comment (KanjiMonster) [reply recipes-extended/onl/onl_git.bb:85]: "vendors for which we build common vendor modules"?

Comment (ideaship) [reply recipes-extended/onl/onl_git.bb:85]: Works for me.

Comment (jklare) [reply recipes-extended/onl/onl_git.bb:85]: WOHOOOOO!!!

Comment (KanjiMonster) [reply recipes-extended/onl/onl_git.bb]: Updated
